### DISCUSSION TITLE
graphql/hive/docs/gateway/other-features/header-propagation

### DISCUSSION THREAD

**Discussion by giscus** [graphql-hive/console#5637] [Docs Discussions]
# graphql/hive/docs/gateway/other-features/header-propagation

Hive Documentation

https://the-guild.dev/graphql/hive/docs/gateway/other-features/header-propagation

**Comment by Warxcell** [comment]
The tricky part is "how to transfer set-cookie headers from subgraphs to client"?
1. `payload.response.headers.get('set-cookie')` returns empty string always.
2. It seems impossible to set multiple headers?

**Reply by EmrysMyrddin** [reply]
Hi ! Do you have a way to show us a reproduction ? It should be working as expected with this configuration:
```ts
import { defineConfig } from '@graphql-hive/gateway'
 
export const gatewayConfig = defineConfig({
  propagateHeaders: {
    fromSubgraphsToClient({ response }) {
      return {
        'set-cookie': response.headers.get('set-cookie')
      }
    }
  }
})
```

**Reply by Warxcell** [reply]
It simply does not.  As stated: ` response.headers.get('set-cookie')` always returns empty string.
```
    fromSubgraphsToClient({response}) {
      console.log(`Set-Cookie "${response.headers.get('set-cookie')}"`)
      return {
        'set-cookie': response.headers.get('set-cookie')
      }
    }
```
![image](https://github.com/user-attachments/assets/19162860-e218-440a-84ec-e9059a4928d0)


This is my fix:

```
import type {GatewayPlugin} from "@graphql-mesh/serve-runtime/typings/types";

export function useForwardCookies(): GatewayPlugin {
  const cookiesByRequest = new WeakMap<Request, string[]>();
  return {
    onFetch(payload) {
      const request = payload.context.request;
      return (payload) => {
        const setCookies: string[] = [];
        payload.response.headers.forEach((val, name) => {
          if (name.toLowerCase() !== 'set-cookie') {
            return
          }
          setCookies.push(val);
        });
        if (setCookies.length) {
          cookiesByRequest.set(request, setCookies)
        }
      }
    },
    onResponse(payload) {
      const cookieStrings = cookiesByRequest.get(payload.request);
      cookieStrings?.forEach(cookieString => {
        payload.response.headers.append('Set-Cookie', cookieString);
      });
    }
  }
}
```

**Reply by ardatan** [reply]
Could you try `payload.response.headers.getSetCookies()` ?

**Reply by Warxcell** [reply]
```TypeError: response.getSetCookies is not a function```

**Reply by ardatan** [reply]
Sorry I was referring to this one;
`response.headers.getSetCookies()`
https://developer.mozilla.org/en-US/docs/Web/API/Headers/getSetCookie

**Reply by Warxcell** [reply]
> Sorry I was referring to this one; `response.headers.getSetCookies()` https://developer.mozilla.org/en-US/docs/Web/API/Headers/getSetCookie

Ah, I see. Didn't knew it existed, but same - it returns empty array :(

![image](https://github.com/user-attachments/assets/1e9c6874-0b0e-46af-9856-017827d55f29)

**Reply by Warxcell** [reply]
Even if I log `response.headers` - they appear empty. Something is deleting them.

![image](https://github.com/user-attachments/assets/78786e88-a82a-497c-a244-82fe98ff88a0)
